File: aide/flow/tracker.py
"""FlowTracker：编排一次 flow 动作（校验 → hooks → 落盘 → git → 输出）。"""

from __future__ import annotations

from pathlib import Path

from aide.core import output
from aide.core.config import ConfigManager
from aide.flow.branch import BranchManager
from aide.flow.errors import FlowError
from aide.flow.git import GitIntegration
from aide.flow.hooks import run_post_commit_hooks, run_pre_commit_hooks
from aide.flow.storage import FlowStorage
from aide.flow.types import FlowStatus, HistoryEntry
from aide.flow.utils import now_iso, now_task_id, normalize_text
from aide.flow.validator import FlowValidator

DEFAULT_PHASES = ["task-optimize", "flow-design", "impl", "verify", "docs", "finish"]


class FlowTracker:
    def __init__(self, root: Path, cfg: ConfigManager):
        self.root = root
        self.cfg = cfg
        self.storage = FlowStorage(root)
        self.git = GitIntegration(root)
        self.branch_mgr = BranchManager(root, self.git, cfg)

    def start(self, phase: str, summary: str) -> bool:
        return self._run(action="start", to_phase=phase, text=summary)

    def next_step(self, summary: str) -> bool:
        return self._run(action="next-step", to_phase=None, text=summary)

    def back_step(self, reason: str) -> bool:
        return self._run(action="back-step", to_phase=None, text=reason)

    def next_part(self, phase: str, summary: str) -> bool:
        return self._run(action="next-part", to_phase=phase, text=summary)

    def back_part(self, phase: str, reason: str) -> bool:
        """返工请求：检测是否已确认，未确认则生成 key。"""
        try:
            self.storage.ensure_ready()

            # 检查是否存在待确认的 back 请求
            if self.storage.has_pending_back_confirm():
                state = self.storage.load_back_confirm_state()
                if state:
                    output.warn("已存在待确认的返工请求")
                    output.info(f"目标环节: {state['target_part']}")
                    output.info(f"原因: {state['reason']}")
                    output.info(f"请执行: aide flow back-confirm --key {state['pending_key']}")
                    return False

            # 生成新的确认 key
            key = self.storage.save_back_confirm_state(phase, reason)
            output.warn("返工需要确认。请先完成以下准备工作:")
            output.info("1. 触发 rework skill 学习返工流程指南")
            output.info("2. 按照指南更新任务文档（记录返工原因和新需求）")
            output.info("3. 完成准备工作后执行:")
            output.info(f"   aide flow back-confirm --key {key}")
            return True

        except FlowError as exc:
            output.err(str(exc))
            return False

    def back_confirm(self, key: str) -> bool:
        """确认返工请求并执行。"""
        try:
            self.storage.ensure_ready()

            # 检查是否存在待确认的请求
            state = self.storage.load_back_confirm_state()
            if state is None:
                output.err("无待确认的返工请求")
                return False

            # 验证 key
            if state.get("pending_key") != key:
                output.err("确认 key 不匹配")
                return False

            # 获取目标阶段和原因
            target_part = state["target_part"]
            reason = state["reason"]

            # 清除确认状态文件
            self.storage.clear_back_confirm_state()

            # 执行实际的 back-part 操作
            result = self._run(action="back-part", to_phase=target_part, text=reason)

            if result:
                output.warn("建议执行 /exit 重新开始对话")

            return result

        except FlowError as exc:
            output.err(str(exc))
            return False

    def issue(self, description: str) -> bool:
        return self._run(action="issue", to_phase=None, text=description)

    def error(self, description: str) -> bool:
        return self._run(action="error", to_phase=None, text=description)

    def clean(self) -> bool:
        """强制清理当前任务

        前提条件：工作区必须干净
        """
        try:
            self.storage.ensure_ready()

            with self.storage.lock():
                status = self.storage.load_status()
                if status is None:
                    output.err("未找到活跃任务，无需清理")
                    return False

                # 执行强制清理
                success, msg = self.branch_mgr.clean_branch_merge()

                if success:
                    output.ok(f"强制清理完成: {msg}")
                else:
                    if "工作区不干净" in msg or "未找到活跃" in msg:
                        output.err(msg)
                    else:
                        # 临时分支情况，需要手动处理
                        output.warn(msg)
                return success

        except FlowError as exc:
            output.err(str(exc))
            return False

    def _run(self, *, action: str, to_phase: str | None, text: str) -> bool:
        try:
            self.storage.ensure_ready()
            config = self.cfg.load_config()
            phases = _get_phases(config)
            validator = FlowValidator(phases)

            normalized_text = normalize_text(text)
            if not normalized_text:
                raise FlowError("文本参数不能为空")

            with self.storage.lock():
                if action == "start":
                    assert to_phase is not None
                    validator.validate_start(to_phase)
                    self.storage.archive_existing_status()

                    # 创建任务分支
                    task_id = now_task_id()
                    task_branch = self.branch_mgr.create_task_branch(
                        task_id=task_id,
                        task_summary=normalized_text,
                    )
                    branch_info = self.branch_mgr.get_active_branch_info()

                    status = FlowStatus(
                        task_id=task_id,
                        current_phase=to_phase,
                        current_step=0,
                        started_at=now_iso(),
                        history=[],
                        source_branch=branch_info.source_branch if branch_info else None,
                        start_commit=branch_info.start_commit if branch_info else None,
                        task_branch=task_branch,
                    )
                    updated, commit_msg = self._apply_action(
                        status=status,
                        action=action,
                        from_phase=None,
                        to_phase=to_phase,
                        text=normalized_text,
                        validator=validator,
                    )
                    # 先保存状态，再执行 git 操作
                    self.storage.save_status(updated)
                    final_status = self._do_git_commit(updated, commit_msg)
                    self.storage.save_status(final_status)
                    output.ok(f"任务开始: {to_phase} (分支: {task_branch})")
                    run_post_commit_hooks(to_phase=to_phase, action=action)
                    return True

                status = self.storage.load_status()
                if status is None:
                    raise FlowError("未找到流程状态，请先运行：aide flow start <环节名> \"<总结>\"")

                current_phase = status.current_phase
                validator.validate_phase_exists(current_phase)

                if action == "next-part":
                    assert to_phase is not None
                    validator.validate_next_part(current_phase, to_phase)
                elif action == "back-part":
                    assert to_phase is not None
                    validator.validate_back_part(current_phase, to_phase)
                else:
                    to_phase = current_phase

                updated, commit_msg = self._apply_action(
                    status=status,
                    action=action,
                    from_phase=current_phase,
                    to_phase=to_phase,
                    text=normalized_text,
                    validator=validator,
                )
                # 先保存状态，再执行 git 操作
                self.storage.save_status(updated)
                final_status = self._do_git_commit(updated, commit_msg)
                self.storage.save_status(final_status)

                # 如果进入 finish 环节，执行分支合并（必须在提交后执行）
                if action == "next-part" and to_phase == "finish":
                    # finish 提交的哈希就是 end_commit
                    finish_commit = None
                    finish_timestamp = None
                    if final_status.history:
                        last_entry = final_status.history[-1]
                        finish_commit = last_entry.git_commit
                        finish_timestamp = last_entry.timestamp

                    success, merge_msg = self.branch_mgr.finish_branch_merge(
                        task_summary=normalized_text,
                        end_commit=finish_commit,
                        finished_at=finish_timestamp,
                    )
                    if not success:
                        output.warn(merge_msg)

                if action == "next-part":
                    output.ok(f"进入环节: {to_phase}")
                elif action == "back-part":
                    output.warn(f"回退到环节: {to_phase}")
                elif action == "error":
                    output.err(f"错误已记录: {normalized_text}")

                run_post_commit_hooks(to_phase=to_phase, action=action)
                return True
        except FlowError as exc:
            output.err(str(exc))
            return False

    def _apply_action(
        self,
        *,
        status: FlowStatus,
        action: str,
        from_phase: str | None,
        to_phase: str,
        text: str,
        validator: FlowValidator,
    ) -> tuple[FlowStatus, str]:
        """应用动作，返回 (更新后的状态, commit消息)。不执行 git 操作。"""
        if action in {"next-part", "back-part"} and from_phase is None:
            raise FlowError("内部错误：缺少 from_phase")

        if action == "next-part":
            assert from_phase is not None
            validator.validate_next_part(from_phase, to_phase)
        elif action == "back-part":
            assert from_phase is not None
            validator.validate_back_part(from_phase, to_phase)
        elif action == "start":
            validator.validate_start(to_phase)
        else:
            validator.validate_phase_exists(to_phase)

        config = self.cfg.load_config()
        run_pre_commit_hooks(
            root=self.root,
            git=self.git,
            status=status,
            from_phase=from_phase,
            to_phase=to_phase,
            action=action,
            config=config,
        )

        message = _build_commit_message(action=action, phase=to_phase, text=text)

        next_step = status.current_step + 1
        entry = HistoryEntry(
            timestamp=now_iso(),
            action=action,
            phase=to_phase,
            step=next_step,
            summary=text,
            git_commit=None,  # 暂时为 None，后续在 git 提交后更新
        )

        history = [*status.history, entry]
        updated_status = FlowStatus(
            task_id=status.task_id,
            current_phase=to_phase,
            current_step=next_step,
            started_at=status.started_at,
            history=history,
            source_branch=status.source_branch,
            start_commit=status.start_commit,
            task_branch=status.task_branch,
        )
        return updated_status, message

    def _do_git_commit(self, status: FlowStatus, message: str) -> FlowStatus:
        """执行 git add + commit，并更新状态中的 commit hash。"""
        self.git.add_all()
        commit_hash = self.git.commit(message)

        # 更新最后一条历史记录的 git_commit
        if status.history:
            last_entry = status.history[-1]
            updated_entry = HistoryEntry(
                timestamp=last_entry.timestamp,
                action=last_entry.action,
                phase=last_entry.phase,
                step=last_entry.step,
                summary=last_entry.summary,
                git_commit=commit_hash,
            )
            updated_history = [*status.history[:-1], updated_entry]
            return FlowStatus(
                task_id=status.task_id,
                current_phase=status.current_phase,
                current_step=status.current_step,
                started_at=status.started_at,
                history=updated_history,
                source_branch=status.source_branch,
                start_commit=status.start_commit,
                task_branch=status.task_branch,
            )
        return status


def _get_phases(config: dict) -> list[str]:
    flow_cfg = config.get("flow", {})
    phases = flow_cfg.get("phases", DEFAULT_PHASES)
    if not isinstance(phases, list) or not phases:
        return DEFAULT_PHASES
    return phases


def _build_commit_message(*, action: str, phase: str, text: str) -> str:
    if action == "issue":
        return f"[aide] {phase} issue: {text}"
    if action == "error":
        return f"[aide] {phase} error: {text}"
    if action == "back-step":
        return f"[aide] {phase} back-step: {text}"
    if action == "back-part":
        return f"[aide] {phase} back-part: {text}"
    return f"[aide] {phase}: {text}"

File: src/flow/tracker.rs
use std::path::Path;

use crate::core::config::{ConfigManager, get_phases};
use crate::core::output;
use crate::flow::branch::BranchManager;
use crate::flow::git::GitIntegration;
use crate::flow::hooks::{run_post_commit_hooks, run_pre_commit_hooks};
use crate::flow::storage::FlowStorage;
use crate::flow::types::{FlowStatus, HistoryEntry};
use crate::flow::validator::FlowValidator;
use crate::utils::{normalize_text, now_iso, now_task_id};

pub struct FlowTracker {
    root: std::path::PathBuf,
    cfg: ConfigManager,
    storage: FlowStorage,
    git: GitIntegration,
}

impl FlowTracker {
    pub fn new(root: &Path, _cfg: &ConfigManager) -> Self {
        Self {
            root: root.to_path_buf(),
            cfg: ConfigManager::new(root),
            storage: FlowStorage::new(root),
            git: GitIntegration::new(root),
        }
    }

    pub fn start(&mut self, phase: &str, summary: &str) -> bool {
        self.run("start", Some(phase), summary)
    }

    pub fn next_step(&mut self, summary: &str) -> bool {
        self.run("next-step", None, summary)
    }

    pub fn back_step(&mut self, reason: &str) -> bool {
        self.run("back-step", None, reason)
    }

    pub fn next_part(&mut self, phase: &str, summary: &str) -> bool {
        self.run("next-part", Some(phase), summary)
    }

    pub fn back_part(&mut self, phase: &str, reason: &str) -> bool {
        match self.do_back_part(phase, reason) {
            Ok(()) => true,
            Err(e) => {
                output::err(&e);
                false
            }
        }
    }

    pub fn back_confirm(&mut self, key: &str) -> bool {
        match self.do_back_confirm(key) {
            Ok(()) => true,
            Err(e) => {
                output::err(&e);
                false
            }
        }
    }

    pub fn issue(&mut self, description: &str) -> bool {
        self.run("issue", None, description)
    }

    pub fn error(&mut self, description: &str) -> bool {
        self.run("error", None, description)
    }

    pub fn clean(&mut self) -> bool {
        match self.do_clean() {
            Ok(()) => true,
            Err(e) => {
                output::err(&e);
                false
            }
        }
    }

    fn do_back_part(&mut self, phase: &str, reason: &str) -> Result<(), String> {
        self.storage.ensure_ready()?;

        if self.storage.has_pending_back_confirm() {
            if let Some(state) = self.storage.load_back_confirm_state()? {
                output::warn("已存在待确认的返工请求");
                output::info(&format!("目标环节: {}", state.target_part));
                output::info(&format!("原因: {}", state.reason));
                output::info(&format!(
                    "请执行: aide flow back-confirm --key {}",
                    state.pending_key
                ));
                return Err(String::new()); // Silent - already printed
            }
        }

        let key = self.storage.save_back_confirm_state(phase, reason)?;
        output::warn("返工需要确认。请先完成以下准备工作:");
        output::info("1. 触发 rework skill 学习返工流程指南");
        output::info("2. 按照指南更新任务文档（记录返工原因和新需求）");
        output::info("3. 完成准备工作后执行:");
        output::info(&format!("   aide flow back-confirm --key {key}"));
        Ok(())
    }

    fn do_back_confirm(&mut self, key: &str) -> Result<(), String> {
        self.storage.ensure_ready()?;

        let state = self.storage.load_back_confirm_state()?;
        let state = match state {
            Some(s) => s,
            None => return Err("无待确认的返工请求".into()),
        };

        if state.pending_key != key {
            return Err("确认 key 不匹配".into());
        }

        let target_part = state.target_part.clone();
        let reason = state.reason.clone();

        self.storage.clear_back_confirm_state()?;

        let result = self.run("back-part", Some(&target_part), &reason);

        if result {
            output::warn("建议执行 /exit 重新开始对话");
        }

        Ok(())
    }

    fn do_clean(&mut self) -> Result<(), String> {
        self.storage.ensure_ready()?;

        let _lock = self.storage.acquire_lock()?;
        let status = self.storage.load_status()?;
        if status.is_none() {
            return Err("未找到活跃任务，无需清理".into());
        }

        let mut branch_mgr = BranchManager::new(&self.root, &self.cfg);
        let (success, msg) = branch_mgr.clean_branch_merge()?;

        if success {
            output::ok(&format!("强制清理完成: {msg}"));
        } else {
            output::warn(&msg);
        }

        Ok(())
    }

    fn run(&mut self, action: &str, to_phase: Option<&str>, text: &str) -> bool {
        match self.do_run(action, to_phase, text) {
            Ok(()) => true,
            Err(e) => {
                if !e.is_empty() {
                    output::err(&e);
                }
                false
            }
        }
    }

    fn do_run(
        &mut self,
        action: &str,
        to_phase: Option<&str>,
        text: &str,
    ) -> Result<(), String> {
        self.storage.ensure_ready()?;
        let config = self.cfg.load_config();
        let phases = get_phases(&config);
        let validator = FlowValidator::new(phases)?;

        let normalized_text = normalize_text(text);
        if normalized_text.is_empty() {
            return Err("文本参数不能为空".into());
        }

        let _lock = self.storage.acquire_lock()?;

        if action == "start" {
            let to_phase = to_phase.ok_or("内部错误：start 缺少 phase")?;
            validator.validate_start(to_phase)?;
            self.storage.archive_existing_status()?;

            let task_id = now_task_id();
            let mut branch_mgr = BranchManager::new(&self.root, &self.cfg);
            let task_branch = branch_mgr.create_task_branch(&task_id, &normalized_text)?;
            let branch_info = branch_mgr.get_active_branch_info()?;

            let status = FlowStatus {
                task_id,
                current_phase: to_phase.to_string(),
                current_step: 0,
                started_at: now_iso(),
                history: Vec::new(),
                source_branch: branch_info.as_ref().map(|b| b.source_branch.clone()),
                start_commit: branch_info.as_ref().map(|b| b.start_commit.clone()),
                task_branch: Some(task_branch.clone()),
            };

            let (updated, commit_msg) = self.apply_action(
                &status,
                action,
                None,
                to_phase,
                &normalized_text,
                &validator,
                &config,
            )?;
            self.storage.save_status(&updated)?;
            let final_status = self.do_git_commit(&updated, &commit_msg)?;
            self.storage.save_status(&final_status)?;
            output::ok(&format!("任务开始: {to_phase} (分支: {task_branch})"));
            run_post_commit_hooks(to_phase, action);
            return Ok(());
        }

        let status = self
            .storage
            .load_status()?
            .ok_or("未找到流程状态，请先运行：aide flow start <环节名> \"<总结>\"")?;

        let current_phase = status.current_phase.clone();
        validator.validate_phase_exists(&current_phase)?;

        let to_phase = match action {
            "next-part" => {
                let tp = to_phase.ok_or("内部错误：next-part 缺少 phase")?;
                validator.validate_next_part(&current_phase, tp)?;
                tp.to_string()
            }
            "back-part" => {
                let tp = to_phase.ok_or("内部错误：back-part 缺少 phase")?;
                validator.validate_back_part(&current_phase, tp)?;
                tp.to_string()
            }
            _ => current_phase.clone(),
        };

        let (updated, commit_msg) = self.apply_action(
            &status,
            action,
            Some(&current_phase),
            &to_phase,
            &normalized_text,
            &validator,
            &config,
        )?;
        self.storage.save_status(&updated)?;
        let final_status = self.do_git_commit(&updated, &commit_msg)?;
        self.storage.save_status(&final_status)?;

        // finish 环节合并
        if action == "next-part" && to_phase == "finish" {
            let finish_commit = final_status
                .history
                .last()
                .and_then(|e| e.git_commit.clone());
            let finish_timestamp = final_status
                .history
                .last()
                .map(|e| e.timestamp.clone());

            let mut branch_mgr = BranchManager::new(&self.root, &self.cfg);
            let (success, merge_msg) = branch_mgr.finish_branch_merge(
                &normalized_text,
                finish_commit.as_deref(),
                finish_timestamp.as_deref(),
            )?;
            if !success {
                output::warn(&merge_msg);
            }
        }

        match action {
            "next-part" => output::ok(&format!("进入环节: {to_phase}")),
            "back-part" => output::warn(&format!("回退到环节: {to_phase}")),
            "error" => output::err(&format!("错误已记录: {normalized_text}")),
            _ => {}
        }

        run_post_commit_hooks(&to_phase, action);
        Ok(())
    }

    fn apply_action(
        &self,
        status: &FlowStatus,
        action: &str,
        from_phase: Option<&str>,
        to_phase: &str,
        text: &str,
        _validator: &FlowValidator,
        config: &toml::Value,
    ) -> Result<(FlowStatus, String), String> {
        // Run pre-commit hooks
        run_pre_commit_hooks(
            &self.root,
            &self.git,
            Some(status),
            from_phase,
            to_phase,
            action,
            config,
        )?;

        let message = build_commit_message(action, to_phase, text);
        let next_step = status.current_step + 1;

        let entry = HistoryEntry {
            timestamp: now_iso(),
            action: action.to_string(),
            phase: to_phase.to_string(),
            step: next_step,
            summary: text.to_string(),
            git_commit: None,
        };

        let mut history = status.history.clone();
        history.push(entry);

        let updated = FlowStatus {
            task_id: status.task_id.clone(),
            current_phase: to_phase.to_string(),
            current_step: next_step,
            started_at: status.started_at.clone(),
            history,
            source_branch: status.source_branch.clone(),
            start_commit: status.start_commit.clone(),
            task_branch: status.task_branch.clone(),
        };

        Ok((updated, message))
    }

    fn do_git_commit(
        &self,
        status: &FlowStatus,
        message: &str,
    ) -> Result<FlowStatus, String> {
        self.git.add_all()?;
        let commit_hash = self.git.commit(message)?;

        if !status.history.is_empty() {
            let mut updated = status.clone();
            if let Some(last) = updated.history.last_mut() {
                last.git_commit = commit_hash;
            }
            return Ok(updated);
        }

        Ok(status.clone())
    }
}

fn build_commit_message(action: &str, phase: &str, text: &str) -> String {
    match action {
        "issue" => format!("[aide] {phase} issue: {text}"),
        "error" => format!("[aide] {phase} error: {text}"),
        "back-step" => format!("[aide] {phase} back-step: {text}"),
        "back-part" => format!("[aide] {phase} back-part: {text}"),
        _ => format!("[aide] {phase}: {text}"),
    }
}
